### PR TITLE
[feature][refactor] Migrate inventory, item. Refactor interact input action.

### LINKED ISSUE #33: Migrate Inventory

### LINKED ISSUE #34: Migrate Item + Flashlight

### LINKED ISSUE #35: Migrate Item Detector
- BoxCollider di kaki player

### LINKED ISSUE #36: Move Interact InputAction to Unity Event

### PR DESCRIPTION
**Pull request by rafiftaris**
### Main changes
- Set interact input action to player events
- Split dark scene into interactable tester and item tester
- Migrated inventory and item class
- Implemented dummy item flashlight for testing

### Tasks List
- [close #33 ]
- [close #34 ]
- [close #35 ]
- [close #36 ]

### Attachments

https://user-images.githubusercontent.com/51603162/148253127-1d6aece7-e80c-48ad-aada-5e941832afbf.mp4

### CODE REVIEW

**Review by naufalhilmi12345 — APPROVED** [review]
LGTM!!

File: Assets/Dev/Scripts/Player/Detectors/ItemDetector.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+[RequireComponent(typeof(BoxCollider))]
+public class ItemDetector : ObjectDetector
+{
+    [Tooltip("The inventory of the player for this item detector")]
+    public Inventory inventory;
+
+    [Tooltip("Detector zone for item")]
+    private BoxCollider detectorZone;
+
+    private void Awake()
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Player/Detectors/ItemDetector.cs:19]: Change to Start() and also change the InteractableDetector as well. Move the GetComponent to the Start() too.

File: Assets/Dev/Scripts/Player/Detectors/ItemDetector.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * Subclass of ObjectDetector class.
+ * Used to detect Item objects with the "PickItem" action.
+ * 
+ * Will call the Inventory to pick this item.
+ */
+[RequireComponent(typeof(BoxCollider))]
+public class ItemDetector : ObjectDetector
+{
+    [Tooltip("The inventory of the player for this item detector")]
+    public Inventory inventory;
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Player/Detectors/ItemDetector.cs:14]: Change to private and add GetComponent in Start()

File: Assets/Dev/Scripts/Utils/MathCalcu.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+* Class that has methods for mathematical calculation
+*/
+public static class MathCalcu
+{
+    public static int mod(int x, int m)
+    {
+        return (x % m + m) % m;
+    }
+}
+}
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Utils/MathCalcu.cs:13]: Add new line

File: Assets/Dev/Scripts/Player/Inventory.cs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    private int activeIdx = 0;
+
+    [Tooltip("The number of item in inventory")]
+    private int numOfItem = 0;
+    #endregion
+
+    #region Variables - Item position adjustment
+    [Space]
+    [Tooltip("Active item height offset")]
+    private Vector3 activeItemYOffset = new Vector3(0, -0.5f, 0);
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Player/Inventory.cs:38]: Change to public and edit it through the inspector

File: Assets/Dev/Scripts/Player/Inventory.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        if (ctx.performed)
+        {
+            if (activeItem)
+            {
+                //Debug.Log("[INVENTORY] Discard " + activeItem.name);
+
+                // Put item as child of Inventory
+                activeItem.SetCollider(true);
+
+                activeItem.gameObject.transform.parent = GameObject.Find("OverworldItems").transform;
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Player/Inventory.cs:112]: make OverworldItems a public var and assign through inspector

File: Assets/Dev/Scripts/Player/Inventory.cs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                    }
+                }
+
+                numOfItem++;
+            }
+            else
+            {
+                items[activeIdx] = item;
+                activeItem = item;
+                numOfItem++;
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Player/Inventory.cs:88]: Add TODO: Update Inventory HUD

File: Assets/Dev/Scripts/Player/Inventory.cs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    }
+    #endregion
+
+    public void UseActiveItem(InputAction.CallbackContext ctx)
+    {
+        if (ctx.performed)
+        {
+            activeItem?.Use();
+            if (!activeItem) Debug.Log("[ITEM] Missing active item");
+        }
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Player/Inventory.cs:161]: Add TODO: Change Active Item HUD

File: Assets/Dev/Scripts/Player/Inventory.cs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    public void DiscardItem(InputAction.CallbackContext ctx)
+    {
+        if (ctx.performed)
+        {
+            if (activeItem)
+            {
+                //Debug.Log("[INVENTORY] Discard " + activeItem.name);
+
+                // Put item as child of Inventory
+                activeItem.SetCollider(true);
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Player/Inventory.cs:110]: move to below SetMeshRenderer

File: Assets/Dev/Scripts/Interactables/Item/Item.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+ */
+[RequireComponent(typeof(MeshRenderer))]
+public abstract class Item : Interactable
+{
+    public abstract void Use();
+
+    public override void OnInteraction()
+    {
+        Pick();
+        SetMeshRenderer(false);
```
Comment (raihansyahdean) [inline Assets/Dev/Scripts/Interactables/Item/Item.cs:16]: Move to Pick()

**Review by raihansyahdean — CHANGES_REQUESTED** [review]
Minor Clean Code

File: Assets/Dev/Scripts/Player/Inventory.cs (hunk 7 above)
Comment (rafiftaris) [reply Assets/Dev/Scripts/Player/Inventory.cs:161]: this function is UseActiveItem, already added in ChangeActiveItem function

**Review by raihansyahdean — APPROVED** [review]
Nicely done!

**Pull request merged**